drivers/net/ethernet/intel/idpf/eth/idpf_eth.c

```diff
@@ -0,0 +1,380 @@
+	if (adapter->vport)
+		idpf_vport_dealloc(adapter->vport);
+	return err;
+}
+
+/**
+ * idpf_eth_probe - Probe ethernet device
+ * @adev: Structure related to auxiliary ethernet device
+ * @id: auxiliary device id
+ */
+static int idpf_eth_probe(struct auxiliary_device *adev,
+			  const struct auxiliary_device_id *id)
+{
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	struct idpf_eth_adapter *adapter;
+	struct pci_dev *pdev;
+	int err;
+
+	eth_dev = container_of(adev, struct idpf_eth_idc_auxiliary_dev, adev);
+	adapter = dev_get_drvdata(&adev->dev);
+	pdev = eth_dev->eth_info.eth_shared->pdev;
+	if (!adapter) {
+		adapter = idpf_eth_adapter_alloc(pdev);
+		if (!adapter)
+			return -ENOMEM;
+
+		/* Initialize dev_info */
+		adapter->dev_info = &eth_dev->eth_info;
+		dev_set_drvdata(&adev->dev, adapter);
+
+	}
+
+	/* Save Eth device info */
+	adapter->dev_info = &eth_dev->eth_info;
+	adapter->dev_info->start_probe_post_init = false;
+
+	/* Device's pre-initialization */
+	err = idpf_eth_device_pre_init(adapter);
+
+	/* Scheduling work for device's post-initialization */
+	queue_delayed_work(adapter->post_init_wq, &adapter->post_init_task,
+			   msecs_to_jiffies(2 * (pdev->devfn & 0x07)));
+	return err;
+}
+
+/**
+ * idpf_eth_shutdown - Shutdown call
+ * @adev: Structure related to auxiliary ethernet device
+ */
+static void idpf_eth_shutdown(struct auxiliary_device *adev)
+{
+}
+
+/**
+ * idpf_eth_remove - removes ethernet device
+ * @adev: Structure related to auxiliary ethernet device
+ */
+static void idpf_eth_remove(struct auxiliary_device *adev)
+{
+	struct idpf_eth_adapter *adapter;
+	struct net_device *netdev;
+
+	adapter = dev_get_drvdata(&adev->dev);
+	if (!adapter)
+		return;
+
+	/* Cancel any pending task(s) */
+	cancel_delayed_work_sync(&adapter->post_init_task);
+	cancel_delayed_work_sync(&adapter->stats_task);
+
+	if (adapter->netdev) {
+		netdev = adapter->netdev;
+		if (netdev) {
+			if (netdev->reg_state != NETREG_UNINITIALIZED)
+				unregister_netdev(netdev);
+			free_netdev(netdev);
+			adapter->netdev = NULL;
+		}
+	}
+
+	kfree(adapter->vport_config);
+	adapter->vport_config = NULL;
+	kfree(adapter->vport);
+	adapter->vport = NULL;
+
+	destroy_workqueue(adapter->post_init_wq);
+	destroy_workqueue(adapter->stats_wq);
+	mutex_destroy(&adapter->vport_ctrl_lock);
+
+	kfree(adapter);
+}
+
+/**
+ * idpf_eth_event_handler - Ethernet device event handler to proess events
+ * @dev_info: Device information
+ * @event: Event info
+ */
+static void idpf_eth_event_handler(struct idpf_eth_idc_dev_info *dev_info,
+				   struct idpf_eth_idc_event *event)
+{
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	struct idpf_eth_adapter *adapter;
+
+	eth_dev = container_of(dev_info, struct idpf_eth_idc_auxiliary_dev,
+			       eth_info);
+	adapter = dev_get_drvdata(&eth_dev->adev.dev);
+	if (!adapter)
+		return;
+
+	switch (event->event_code) {
+	case IDPF_ETH_IDC_EVENT_RESET_INITIATED:
+		set_bit(IDPF_ETH_RESET_IN_PROG, adapter->flags);
```

Comment:
just bring this set_bit inside the mutex lock block

Reply:
done !